Async BAM performance improvements

d-cameron opened the pull request:
### Description

The existing aysnc read code performs read-ahead and decompression on a seperate thread. Combining with AsyncBufferedIterator and we can get ~3x performance increase when doing light processing (eg. picard CollectInsertSizeMetrics). This PR increases the level of parallelism to close to linear with the number of cores (tested to ~16x) through the following changes:
- Separation of async BlockCompressedInputStream read-ahead and decompression into separate tasks to enable parallel decompression of multiple gzip blocks
- Parallel async decoding of BAM records.
  - whilst this incurs the overhead of an additional buffer copy, it enables SAMRecord decoding to be performed in parallel
  - By default, SAMRecords are decoded in 8KiB batches to prevent excessive thread synchronisation overhead when processing small SAM records. 

The default 128k buffer size results in read-ahead of only 2 gzip block thus only 2 concurrent decompression tasks. Specifying a larger buffer size (e.g. ` -Dsamjdk.buffer_size=4194304`) enables scaling to additional cores.

### Checklist

- [ X ] Code compiles correctly
- [ X ] New tests covering changes and new functionality
- [ X ] All tests passing (tested on windows so some unrelated test cases fail)
- [ NA ] Extended the README / documentation, if necessary
- [  ] Is not backward compatible (breaks binary or source compatibility)

lbergelson commented:
@d-cameron Thank you for this PR.  It sounds like a big speedup for bam reading.  

It's fairly complicated multithreaded code in a mission critical section of the codebase so it might take some time to review.

yfarjoun commented:
Hey @d-cameron.

This seems like a really useful feature for htsjdk. A few of us have read through and were wondering why you didn't use CompletableFutures as the underlying synchronization pattern. 

Using this pattern would simplify the review process, further concerns about possible locks and eliminate the need for explicit sleeps in the code. In addition, Expections will be propagated across threads helping in error messaging and debugging. 

Would you be amenable to modifying your code to use CompletableFutures? We understand the uncertainty around the process of getting a PR merged, and are committed to working with you towards that goal.

d-cameron commented:
Switching to CompleteableFutures will remove some of the sleeps, but not all. Note that even with CompleteableFutures, the code still needs a review for deadlocks and race conditions, although the complexity would mostly be limited to just the async read task.

The issue with the async read task is that the next read task can be scheduled either by the foreground thread, or the background thread and at most one of these tasks can be running at any point. We can't just have a single long-running task as that can deadlock when the thread pool is limited and results in unfair read-ahead scheduling when there are multiple files being read in parallel. Since the task can be scheduled 

I've thought of an alternate design in which the read-ahead task are all scheduled from the calling thread so I'll rewrite using that approach with CompleteableFutures and see how much cleaner it ends up being.

yfarjoun commented:
much appreciated.

d-cameron commented:
@yfarjoun @lbergelson refactored so that all background processing tasks are CompletableFutures scheduled from the foreground thread. This version should be much simpler to code review. I've split the refactoring to two separate commits - one with the CompletableFuture refactor, and a second adding the `RecordOrException` complexity required for exceptions to be raised at the same time as they are raised when reading synchronously

lbergelson commented:
@d-cameron Thank you!  That's great.  We will review this this week.

cmnbroad commented:
Hi @d-cameron It looks like this is a couple of months behind master - can you rebase on current master before we start reviewing ? Thanks.

lbergelson commented:
@d-cameron I'm sorry I haven't gotten through reviewing this yet.  I expected to get an htsjdk release out this week and then review it after that, but I've run into a number of blocking issues that ate my time and I haven't gotten to give this a proper look yet.  It is one of my priorities to get to next week though.  It will not be abandoned forever this time...

lbergelson commented on src/main/java/htsjdk/samtools/BAMFileReader.java:
```diff
@@ -796,59 +804,160 @@ public SAMRecord next() {
+                } else {
+                    if (useAsynchronousIO) {
+                        byte[] buffer = new byte[readLength];
+                        mCompressedInputStream.read(buffer);
```
There's already a method in BlockCompressedInputStream that does this called readBytes but it's a private static method.  (it's also duplicated 4 times in different places in our code base...) It might make sense to expose it as a public method in IOUtil and and then use that here instead of read.

lbergelson commented on src/main/java/htsjdk/samtools/BAMFileReader.java:
```diff
@@ -772,9 +772,11 @@ public SAMRecord next() {
         private final BAMRecordCodec bamRecordCodec;
-        private long samRecordIndex = 0; // Records at what position (counted in records) we are at in the file
+        private final AsyncBamDecoder mAsync;
+        private final ConcurrentLinkedQueue<BAMRecordCodec> mAvailableIdleCodec;
```
I poked at this a bit because I don't like having separate `BAMRecordCodec` and `mAvailableIdleCodec` fields.  I think it can be cleaned up by using an ObjectPool<BAMRecordCodec>.
Here's a simple implementation that I believe works (although I haven't tested it yet).

```
public interface ObjectPool<T> {
    <U> U borrow(Function<T, U> computeWithPoolObject);
}

public class QueuePool<T> implements ObjectPool<T> {
    private final Queue<T> queue = new ConcurrentLinkedQueue<>();
    private final Supplier<T> objectSupplier;

    public QueuePool(Supplier<T> objectSupplier) {
        this.objectSupplier = objectSupplier;
    }

    @Override
    public <U> U borrow(Function<T,U> usePoolObject) {
        T object = queue.poll();
        try {
            if(object == null){
                object = objectSupplier.get();
            }
            return usePoolObject.apply(object);
        } finally {
            queue.offer(object);
        }
    }
}

public class SingletonPool<T> implements ObjectPool<T> {
    private final T object;

    SingletonPool(T object){
        this.object = object;
    }

    @Override
    public synchronized <U> U borrow(final Function<T, U> usePoolObject) {
        return usePoolObject.apply(object);
    }
}
```

The two fields can be merged into a single ObjectPool that's instantiated either as a singleton or a queue. 

I think there are a bunch of places in BlockCompressedInputStream and AsyncBlockCompressedInputStream that could be significantly simplified using something like this as well.  It may need more complicated new object generation though, I haven't worked through it yet.

d-cameron replied:
> I don't like having separate BAMRecordCodec and mAvailableIdleCodec fields.

bamRecordCodec is the codec attached to the stream and is the only codec used in sync calls. In async mode, the stream codec is used to read the record length from the stream, and the available codecs are recycled for parallel record parsing. They're both needed because they do different things.

One could add the stream codec to the pool and reattached to the stream every time you needed to read the record length but that's even less intuitive than what I have now. I've renamed the variables so it's clear what they do.

d-cameron replied:
>I think there are a bunch of places in BlockCompressedInputStream and AsyncBlockCompressedInputStream that could be significantly simplified using something like this as well. It may need more complicated new object generation though, I haven't worked through it yet.

How do you like my RecyclingSupplier implementation? AsyncBlockCompressedInputStream.mFreeDecompressedBlockedBuffers doesn't fit the model due to the potential for different buffer sizes so that's been kept as a ConcurrentLinkedQueue but the others have been simplified.

d-cameron replied:
Decided to just remove the codec cache entirely. BamRecordCodec is pretty cheap to clone and it removes a per record (not per batch) source of synchronisation

lbergelson reviewed (changes requested):
@d-cameron I'm not finished reviewing this yet, but I thought I'd post my initial thoughts so you can see that we're not totally ignoring you.  I had to make an unplanned trip last week that ate into my review time.

Generally this looks much nicer to me than the previous version.  

There are few things so far that I think could be made clearer.  The buffer/ decoder leasing is confusing and adds a lot of lines of code right now.  (especially in the BlockCompressedInputStream classes.)  I think it could be cleaned up by adding an explicit notion of ObjectPools.  That might also enable the two codepaths to be unified more with less special casing of things.   I'll finish taking a look tomorrow.  

There's a lot of little cleanup things that could be done, making things more things final, adding new lines between methods, etc, that I haven't labelled every instance of.  I don't mind doing my own pass on that sort of thing after we work out the details of this.

d-cameron commented on src/main/java/htsjdk/samtools/BAMFileReader.java:
```diff
@@ -796,59 +804,160 @@ public SAMRecord next() {
+            private Long stop;
+            private final long recordIndex;
+            private SAMRecord record;
+            public BamRecordDecodingInfo(long recordIndex, int recordLength, long start) {
```
I've taking a slightly different approach delegated in the other direction. I feel Long is safer as it's absolute clear that there is no stop yet thus making it impossible for future changes to accidentally take a file pointer a the negative placeholder value.

d-cameron commented on src/main/java/htsjdk/samtools/util/AsyncReadTaskRunner.java:
```diff
@@ -0,0 +1,274 @@
+ */
+public abstract class AsyncReadTaskRunner<T, U> {
+    private static final Log log = Log.getInstance(AsyncReadTaskRunner.class);
+    private static Executor nonblockingThreadpool = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors(), r -> {
```
Added to Defaults

d-cameron commented on src/main/java/htsjdk/samtools/util/AsyncReadTaskRunner.java:
```diff
@@ -0,0 +1,274 @@
+     * as possible.
+     */
+    public static Executor getBlockingThreadpool() { return blockingThreadpool; }
+    public static void setNonblockingThreadpool(Executor nonblockingThreadpool) {
```
As it doesn't really matter much which thread pool tasks get scheduled to I went with the simple/conservative approach that existing tasks were free to complete on the old thread pool, with new ones getting scheduled on the new one. This approach also avoids having to synchronise on the thread pool object as well as avoiding having to block/throw in the set() method.

Since the operations in progress are free to complete on the existing thread pool, there's no need to throw.

d-cameron commented on src/main/java/htsjdk/samtools/util/AsyncReadTaskRunner.java:
```diff
@@ -0,0 +1,274 @@
+ * and any number of non-blocking transform tasks (transform()).
+ * Synchronisation of the underlying stream/iterator is handled by ensuring that performReadAhead() is not executed
+ * by more than one thread concurrently.
+ * As this represents a serial bottleneck limiting palatalisation,
```
parallelization

d-cameron commented on src/main/java/htsjdk/samtools/util/AsyncBlockCompressedInputStream.java:
```diff
@@ -32,203 +32,120 @@
+    private final ConcurrentLinkedQueue<BlockGunzipper> mFreeInflators = new ConcurrentLinkedQueue<>();
+    private final AsyncBlockCompressedInputStreamTaskRunner async = new AsyncBlockCompressedInputStreamTaskRunner(
+            BlockCompressedStreamConstants.MAX_COMPRESSED_BLOCK_SIZE,
+            Math.max(1, Defaults.NON_ZERO_BUFFER_SIZE / BlockCompressedStreamConstants.MAX_COMPRESSED_BLOCK_SIZE));
```
Should htsjdk err on the side of too big or too small buffers? If someone asks for 100KiB of buffer, should htsjdk give them 1 or 2 64KiB buffers? Either option is defensible so I though it'd be best to defer to you.

d-cameron commented on src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java:
```diff
@@ -712,38 +670,61 @@ private static boolean preambleEqual(final byte[] preamble, final byte[] buf, fi
         }
         return true;
     }
+    protected class CompressionBlock {
```
I went with static and passing in the BlockCompressedInputStream as a parameter as this approach avoids unnecessary calls to getSource() (which is only invoked when there is an error).

d-cameron commented on src/main/java/htsjdk/samtools/util/AsyncReadTaskRunner.java:
```diff
@@ -0,0 +1,274 @@
+        return t;
+    });
+
+    private static Executor blockingThreadpool = Executors.newCachedThreadPool(r -> {
```
Correct. This class ensures that we don't attempt to read from the underlying stream on multiplt threads at one.

This approach reduces the number of threads created to the number of read-ahead in flight at any one time. Generally speaking, this will be much lower count when you have many files open. A private thread per file introduces an overhead of over 1MB (default stack size) per file and that memory is better used in actual read-ahead buffers. We don't want to spin up OS threads unless they're actually going to be used.

The shared pool also allows the scheduling of these tasks to configured by the end user by replacing the executor.

d-cameron replied:
Unless you're pulling your bams from the internet, the bottleneck will be in the decompression and parsing. In my tracing, I've only ever got up to a handful of blocking threads even when processing 100+ files in parallel.

d-cameron commented on src/main/java/htsjdk/samtools/util/AsyncReadTaskRunner.java:
```diff
@@ -0,0 +1,274 @@
+     * The results of these tasks are discarded and any exceptions raised
+     * during processing of these tasks are swallowed.
+     */
+    public void flushAsyncProcessing() {
```
Waiting for completion is the safest way to ensure that the underlying stream is in a valid state and that all outstanding tasks have actually completed.

That said, we don't actually need to complete all the tasks, we just need to not interrupt the underlying operations. I've added a short-circuit to safely stop processing.

d-cameron commented on src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java:
```diff
@@ -455,100 +452,61 @@ private static boolean isValidBlockHeader(final byte[] buffer) {
-        mCurrentBlock = nextBlock(getBufferForReuse(mCurrentBlock));
+        mCurrentBlock = nextBlock(mCurrentBlock);
+        if (mCurrentBlock == null) {
+            // sentinel EOF block
```
Unfortunately not. mStreamOffset isn't known before reaching EOF.

d-cameron commented:
Back to you @lbergelson

d-cameron commented:
> Extended the README / documentation, if necessary

Is there a README / documentation with all the settings exposed through `Defaults.java`? This PR now exposes the threadpool size through a new one.

d-cameron commented:
@lbergelson @yfarjoun Any update on this PR?

yfarjoun commented:
@lbergelson please review this when you have time.

lbergelson commented:
Oh no.  The PR that I swore to review quickly and then never did.    @d-cameron It looks like there are build failures due to a missing logger and a typo.  Inflater/inflator strikes again.  

I uploaded a fix here https://github.com/samtools/htsjdk/tree/lb_async_bam_performance_improvements

d-cameron commented:
Fixes in; updated to latest master

Commit:
Using separate async tasks for IO and computation for both gzip decompression and BAM record parsing

Commit:
Corrected incorrectly failing test case
Refactored AsyncReadTaskRunner test case to handle our early abort redesign.
Now need to test against the threadpool itself as we will now skip any scheduled read-aheads of transforms that we haven't already sent to the thread pool.

Commit:
Added RecyclingSupplier test coverage

Commit:
Ensuring all test classes extend HtsJdkTest

Commit:
Optimised imports; fixed spelling of HtsjdkTest

Commit:
Removed AsyncReadTaskRunnerTest logging spam

Commit:
Updated test case to handle possibility that not all read tasks were complete after the sleep

Commit:
Reduced lock contention by only attempting job scheduling when buffer is empty

jmarshall commented:
I have taken the liberty of rebasing this onto current **master**, to resolve the (trivial) merge conflict and rerun the CI tests.

jmarshall committed:
Put assertEquals() actual/expected the right way around [minor]

jmarshall commented:
I have been unable to reproduce the `testDisableAsyncProcessingLetsExistingTasksComplete` failure locally with OpenJDK 8 (java-1.8.0-openjdk on CentOS 8). Looking at the difference between builds [6577](https://app.travis-ci.com/github/samtools/htsjdk/builds/235643417) and [6578](https://app.travis-ci.com/github/samtools/htsjdk/builds/235652793) it appears that the issue is intermittent… (the worst kind!)
